File: go/signing/secret.go
/*
Package signing implements Standard Webhooks (https://www.standardwebhooks.com)
HMAC-SHA256 signature generation and verification. It mirrors packages/signing
in the TypeScript implementation, which remains the source of truth for
behavior and wire format.

Sign/Verify are byte-for-byte compatible with the official standardwebhooks
libraries and with the TS @anyhook/signing package -- see the golden vector
test in sign_test.go.
*/
package signing

import (
	"crypto/rand"
	"encoding/base64"
	"strings"
)

// SecretPrefix is prepended to generated signing secrets.
const SecretPrefix = "whsec_"

// DefaultSecretBytes is the number of random bytes GenerateSecret encodes by default.
const DefaultSecretBytes = 24

// GenerateSecret generates a new Standard Webhooks-compatible signing secret:
// a whsec_-prefixed, base64-encoded random byte string. Pass 0 for the
// default (24 bytes).
func GenerateSecret(numBytes int) (string, error) {
	if numBytes <= 0 {
		numBytes = DefaultSecretBytes
	}
	b := make([]byte, numBytes)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	return SecretPrefix + base64.StdEncoding.EncodeToString(b), nil
}

// ParseSecret parses a webhook signing secret into raw HMAC key bytes.
// Strips the whsec_ prefix (if present) and base64-decodes the remainder.
func ParseSecret(secret string) ([]byte, error) {
	raw := strings.TrimPrefix(secret, SecretPrefix)
	return base64.StdEncoding.DecodeString(raw)
}

File: packages/signing/src/secret.ts
/** @fileoverview Webhook signing secret generation and parsing. @module @anyhook/signing */

const SECRET_PREFIX = 'whsec_';

/**
 * Generates a new Standard Webhooks-compatible signing secret.
 * @param bytes - Number of random bytes to encode (default 24)
 * @returns A `whsec_`-prefixed base64-encoded secret
 */
export function generateSecret(bytes = 24): string {
  const random = new Uint8Array(bytes);
  crypto.getRandomValues(random);
  return `${SECRET_PREFIX}${Buffer.from(random).toString('base64')}`;
}

/**
 * Parses a webhook signing secret into raw HMAC key bytes.
 * Strips the `whsec_` prefix (if present) and base64-decodes the remainder.
 * @param secret - The secret string, with or without the `whsec_` prefix
 * @returns The decoded key bytes
 */
export function parseSecret(secret: string): Uint8Array {
  const raw = secret.startsWith(SECRET_PREFIX) ? secret.slice(SECRET_PREFIX.length) : secret;
  return new Uint8Array(Buffer.from(raw, 'base64'));
}
